main/src/ca/uwaterloo/flix/api/lsp/provider/RenameProvider.scala

```diff
@@ -27,145 +30,517 @@ object RenameProvider {
-    val defn = sym.loc
-    val uses = index.usesOf(sym)
-    rename(newName, uses + defn)
+  private def getStructSymOccurs(sym: Symbol.StructSym)(implicit root: Root): Set[SourceLocation] = {
+    var occurs: Set[SourceLocation] = Set.empty
+    def consider(s: Symbol.StructSym, loc: SourceLocation): Unit = {
+      if (s == sym) { occurs += loc }
+    }
+    object StructSymConsumer extends Consumer {
+      override def consumeStruct(struct: TypedAst.Struct): Unit = consider(struct.sym, struct.sym.loc)
+      override def consumeType(tpe: Type): Unit = tpe match {
+        case Type.Cst(TypeConstructor.Struct(sym, _), loc) => consider(sym, loc)
+        case _ => ()
+      }
+    }
+
+    Visitor.visitRoot(root, StructSymConsumer, AllAcceptor)
+
+    occurs
   }
 
-  private def renameLabel(label: Name.Label, newName: String)(implicit index: Index): JObject = {
-    val defs = index.defsOf(label)
-    val uses = index.usesOf(label)
-    rename(newName, defs ++ uses)
+  private def getEnumSymOccurs(sym: Symbol.EnumSym)(implicit root: Root): Set[SourceLocation] = {
+    var occurs: Set[SourceLocation] = Set.empty
+    def consider(s: Symbol.EnumSym, loc: SourceLocation): Unit = {
+      if (s == sym) { occurs += loc }
+    }
+    object EnumSymConsumer extends Consumer {
+      override def consumeEnum(enm: TypedAst.Enum): Unit = consider(enm.sym, enm.sym.loc)
+      override def consumeCaseSymUse(sym: SymUse.CaseSymUse): Unit = seperateEnumAndCaseSymOccur(sym)._1.foreach(enmLoc => consider(sym.sym.enumSym, enmLoc))
+
+      override def consumeType(tpe: Type): Unit = tpe match {
+        case Type.Cst(TypeConstructor.Enum(sym, _), loc) => consider(sym, loc)
+        case _ => ()
+      }
+
+    }
+
+    Visitor.visitRoot(root, EnumSymConsumer, AllAcceptor)
+
+    occurs
   }
 
-  private def renamePred(pred: Name.Pred, newName: String)(implicit index: Index): JObject = {
-    val defs = index.defsOf(pred)
-    val uses = index.usesOf(pred)
-    rename(newName, defs ++ uses)
+  private def getCaseSymOccurs(sym: Symbol.CaseSym)(implicit root: Root): Set[SourceLocation] = {
+    var occurs: Set[SourceLocation] = Set.empty
+    def consider(s: Symbol.CaseSym, loc: SourceLocation): Unit = {
+      if (s == sym) { occurs += loc }
+    }
+    object CaseSymConsumer extends Consumer {
+      override def consumeCase(cse: TypedAst.Case): Unit = consider(cse.sym, cse.sym.loc)
+      override def consumeCaseSymUse(sym: SymUse.CaseSymUse): Unit = consider(sym.sym, seperateEnumAndCaseSymOccur(sym)._2)
+    }
+
+    Visitor.visitRoot(root, CaseSymConsumer, AllAcceptor)
+
+    occurs
   }
 
-  private def renameCase(sym: Symbol.CaseSym, newName: String)(implicit index: Index): JObject = {
-    val defn = sym.loc
-    val uses = index.usesOf(sym)
-    rename(newName, uses + defn)
+  private def seperateEnumAndCaseSymOccur(use: SymUse.CaseSymUse): (Option[SourceLocation], SourceLocation) = {
```

Comment:
This is not a good solution.

Reply:
We should leave it broken (as it already is), or disable the feature, or implement it correctly.